packages/start/vite/plugin.js

```diff
@@ -671,44 +672,48 @@ export default function solidStart(options) {
   ].filter(Boolean);
 }
 
+/**
+ * @returns {import('vite').Plugin}
+ */
 function islands() {
   return {
     name: "solid-start-islands",
-    load(id) {
-      if (id.endsWith("?island")) {
-        return {
-          code: `
-            import Component from '${id.replace("?island", "")}';
+    async transform(code, id, opts) {
+      if (code.includes("unstable_island")) {
+        const ssr = opts?.ssr === true;
 
-            window._$HY.island("${id.slice(process.cwd().length)}", Component);
+        /** @type {Record<string, string>} */
+        const islandComponentIds = Object.fromEntries(
+          await Promise.all(
+            [
+              ...code.matchAll(
+                /const ([A-Za-z_]+) = unstable_island\(\(\) => import\((("([^"]+)")|('([^']+)'))\)\)/g
+              )
+            ].map(async ([a, b, c]) => {
+              c = c.slice(1, -1);
+              const resolved = await this.resolve(c, id);
+              if (!resolved) {
+                throw new Error(`Cannot resolve ${c} from ${id}`);
+              }
+              const componentId = createHash(resolved.id);
+              return [c, componentId];
+            })
+          )
+        );
 
-            export default Component;
-            `
-        };
-      }
-    },
-    /**
-     * @param {any} id
-     * @param {string} code
-     */
-    transform(code, id, ssr) {
-      if (code.includes("unstable_island")) {
         let replaced = code.replaceAll(
           /const ([A-Za-z_]+) = unstable_island\(\(\) => import\((("([^"]+)")|('([^']+)'))\)\)/g,
           (a, b, c) => {
             c = c.slice(1, -1);
+            const componentId = islandComponentIds[c];
             return ssr
-              ? `import ${b}_island from "${c}";
-                  const ${b} = unstable_island(${b}_island, "${
-                  join(dirname(id), c)
-                    .slice(process.cwd().length + 1)
-                    .replaceAll("\\", "/") +
-                  ".tsx" +
-                  "?island"
-                }");`
-              : `const ${b} = unstable_island(() => import("${c}?island"), "${
-                  join(dirname(id), c).replaceAll("\\", "/") + ".tsx" + "?island"
-                }")`;
+              ? `import ${b}_island from ${JSON.stringify(c)};
+                  const ${b} = unstable_island(${b}_island, ${JSON.stringify(componentId)});`
+              : `const ${b}_island_loader = () => import(${JSON.stringify(c)});
```

Comment:
`() => import()` will be transformed to `() => __vitePreload(() => import(), [/* depending modules */])` by Vite later.
This will enable [the async-chunk loading optimization feature](https://vitejs.dev/guide/features.html#async-chunk-loading-optimization).

Before this PR, `hydrateIslands` was using `import(/* @vite-ignore */ asset)` and wasn't able to use this feature. (Because Vite won't be able to analyze that dynamic import.)